### DISCUSSION TITLE
How to add letters that have a custom meaning in the nvim tree views (before, after, signcolumn, etc) ?

### DISCUSSION THREAD

**Discussion by DMyashkov** [nvim-tree/nvim-tree.lua#2963] [Q&A]
I want to integrate harpoon into nvim tree - i have shortcuts alt + a s q w z x for navigating trough files with harpoon(I've already written a function which creates a table - file to letter) I want to add the letter or icon representing where in the harpoon list the file is in any of these: 

`renderer.icons.*_placement` options may be:
- `"before"` : before file/folder, after the file/folders icons
- `"after"` : after file/folder
- `"signcolumn"` : far left, requires |nvim-tree.view.signcolumn| enabled
- `"right_align"` : far right
but i can't seem to place custom icons or letters in any of these places. 

If anyone knows I would also like to add a border to the files that i want to mark.
Here's the harpoon function
```
local function getHarpoonList()
	local harpoon = require("harpoon")
	local mapping = {}
	local harpoon_list = harpoon:list()

	local letters = { "a", "s", "q", "w", "z", "x" }
	for index, item in ipairs(harpoon_list) do
		if index <= #letters then
			mapping[item.value] = letters[index]
		end
	end

	return mapping
end
local mappings = getHarpoonList()
```

**Comment by alex-courtis** [comment]
That sounds really interesting.

This is a good case for a custom decorator #2948 

Your decorator could supply
1. icon or letter for signcolumn / before etc.
2. highlight group

Unfortunately this support has not yet been implented: we need a mechanism to add the decorator and the base class it should extend.

**Comment by DMyashkov** [comment]
@alex-courtis Thanks a lot for the reply when trying to achieve this I wrote this code into the bookmarks.lua file:
```
local HL_POSITION = require("nvim-tree.enum").HL_POSITION
local ICON_PLACEMENT = require("nvim-tree.enum").ICON_PLACEMENT

local Decorator = require("nvim-tree.renderer.decorator")

---@class (exact) DecoratorHarpoon: Decorator
---@field icon HighlightedString?
local DecoratorHarpoon = Decorator:new()

local function getHarpoonList()
  local harpoon = require("harpoon")
  local mapping = {}
  local harpoon_list = harpoon:list().items
  -- vim.notify(vim.inspect(harpoon_list))

  local letters = { "a", "s", "q", "w", "z", "x" }
  for index, item in ipairs(harpoon_list) do
    if index <= #letters then
      local absolute_path = vim.fn.fnamemodify(item.value, ":p")
      mapping[absolute_path] = letters[index]
    end
  end

  --paste mapping
  -- vim.notify(vim.inspect(mapping))

  return mapping
end

local mappings = getHarpoonList()

---@param node Node
local function is_harpoon_item(node)
  return mappings[node.absolute_path] ~= nil
end

---Static factory method
---@param opts table
---@param explorer Explorer
---@return DecoratorHarpoon
function DecoratorHarpoon:create(opts, explorer)
  ---@type DecoratorHarpoon
  local o = {
    explorer = explorer,
    enabled = true,
    hl_pos = HL_POSITION[opts.renderer.highlight_bookmarks] or HL_POSITION.none,
    icon_placement = ICON_PLACEMENT[opts.renderer.icons.bookmarks_placement] or ICON_PLACEMENT.none,
  }
  o = self:new(o) --[[@as DecoratorHarpoon]]

  o.icon = {
    str = "a",
    hl = { "IncSearch" },
  }
  o:define_sign(o.icon)
  return o
end

---Bookmark icon: renderer.icons.show.bookmarks and node is marked
---@param node Node
---@return HighlightedString[]|nil icons

function DecoratorHarpoon:calculate_icons(node)
  local harpoon_letter = mappings[node.absolute_path]
  if harpoon_letter then
    vim.notify("Harpoon letter: " .. harpoon_letter)
    self.icon = {
      str = harpoon_letter,
      hl = { "Search" },
    }
    self:define_sign(self.icon)
    return { self.icon }
  end
end

---Bookmark highlight: renderer.highlight_bookmarks and node is marked
---@param node Node
---@return string|nil group
function DecoratorHarpoon:calculate_highlight(node)
  if is_harpoon_item(node) then
    return "IncSearch"
  end
end

return DecoratorHarpoon


```
the problem is i struggle to make the icon/text different on each file - a k a different letters for the different shortcuts

I found that if i do

```
function DecoratorHarpoon:calculate_icons(node)
  local harpoon_letter = mappings[node.absolute_path]
  if harpoon_letter then
    vim.notify("Harpoon letter: " .. harpoon_letter)
    self.icon = {
      str = harpoon_letter,
      hl = { "Search" },
    }
    self:define_sign(self.icon)
    return { self.icon }
  end
end
```
all of the marks get set to "a" or all to "s" even though the vim notify shows that their all different (a s q w) it turns out that it just sets all of the decorators to the last decorator in the nvim tree ( a k a the marked harpoon file that is most further down in the tree). Seems like self is shared.  If i don't do self:define_sign(self.icon) and try to return something other than self.icon - if it is a different hl group it throws an error that it doesnt recognise it if it's another letter it writes the letter that's written in the constructor and not the new one.

Do you have any idea how i can bypass that and achieve what i want?

**Reply by alex-courtis** [reply]
Thanks for showing me this - it will greatly assist in building #2948 as your requirements are now known.

`Decorator` is not clear on how it works, its lifecycle and what is needed, especially the sign definitions. It also lacks the ability to dynamically (re)define signs. That will be addressed in #2948 

Initial thoughts:
* It looks like `getHarpoonList` is only called once when the explorer is created and stored in a local, hence mappings will be out of date and incorrect
* `define_sign` is called many times, creating or replacing a sign; it should only be done once. It's not  clear how this _should_ work and the expected results.

Please send detailed instructions:
* how to install and use harpoon - assume NO knowledge of what harpoon is or how it works)
* how this decorator should behave - steps to take to use this decorator
We can then work on this together.

**Comment by DMyashkov** [comment]
@alex-courtis Thank you for being so helpful and compliant

https://github.com/ThePrimeagen/harpoon/tree/harpoon2 - Here's the gitub repo for harpoon. Here's my setup and installation with lazy.lua
```
return {
	"ThePrimeagen/harpoon",
	branch = "harpoon2",
	dependencies = { "nvim-lua/plenary.nvim", "nvim-telescope/telescope.nvim" },
	config = function()
		local harpoon = require("harpoon")
		harpoon:setup({
			settings = {
				save_on_toggle = true,
				-- sync_on_ui_close = true,
			},
		})

		-- basic telescope configuration
		local conf = require("telescope.config").values
		local function toggle_telescope(harpoon_files)
			local file_paths = {}
			for _, item in ipairs(harpoon_files.items) do
				table.insert(file_paths, item.value)
			end

			require("telescope.pickers")
				.new({}, {
					prompt_title = "Harpoon",
					finder = require("telescope.finders").new_table({
						results = file_paths,
					}),
					previewer = conf.file_previewer({}),
					sorter = conf.generic_sorter({}),
				})
				:find()
		end

		vim.keymap.set("n", "<leader>ht", function()
			toggle_telescope(harpoon:list())
		end, { desc = "Open harpoon menu with telescope" })

		vim.keymap.set("n", "<leader>ha", function()
			harpoon:list():add()
		end, { desc = "Add current file to harpoon" })
		vim.keymap.set("n", "<leader>hh", function()
			harpoon.ui:toggle_quick_menu(harpoon:list())
		end, { desc = "Toggle harpoon mini menu" })
		vim.keymap.set("n", "<leader>hx", function()
			harpoon:list():remove()
		end, { desc = "Delete current file from harpoon" })
		vim.keymap.set({ "n", "i", "v" }, "<M-a>", function()
			harpoon:list():select(1)
		end, { desc = "Navigate to file 1 in harpoon" })
		vim.keymap.set({ "n", "i", "v" }, "<M-s>", function()
			harpoon:list():select(2)
		end, { desc = "Navigate to file 2 in harpoon" })
		vim.keymap.set({ "n", "i", "v" }, "<M-q>", function()
			harpoon:list():select(3)
		end, { desc = "Navigate to file 3 in harpoon" })
		vim.keymap.set({ "n", "i", "v" }, "<M-w>", function()
			harpoon:list():select(4)
		end, { desc = "Navigate to file 4 in harpoon" })
		vim.keymap.set({ "n", "i", "v" }, "<M-z>", function()
			harpoon:list():select(5)
		end, { desc = "Navigate to file 5 in harpoon" })
		vim.keymap.set({ "n", "i", "v" }, "<M-x>", function()
			harpoon:list():select(6)
		end, { desc = "Navigate to file 6 in harpoon" })
	end,
}

```
Basically it's a plugin made for faster navigation where you have a buffer/list and with harpoon:list():add() you add the current file you are in to a series of relative paths(relative to project root) which would look something like this:


To open this buffer which you can edit with all the same vim shortcuts I have set the shortcut
```
vim.keymap.set("n", "<leader>hh", function()
	harpoon.ui:toggle_quick_menu(harpoon:list())
end, { desc = "Toggle harpoon mini menu" })
```
you can get the contents of the list with harpoon:list() as I have done in the getHarpoonList function
```
local function getHarpoonList()
  local harpoon = require("harpoon")
  local mapping = {}
  local harpoon_list = harpoon:list().items
  -- vim.notify(vim.inspect(harpoon_list))

  local letters = { "a", "s", "q", "w", "z", "x" }
  for index, item in ipairs(harpoon_list) do
    if index <= #letters then
      local absolute_path = vim.fn.fnamemodify(item.value, ":p")
      mapping[absolute_path] = letters[index]
    end
  end

  --paste mapping
  -- vim.notify(vim.inspect(mapping))

  return mapping
end
```
In my setup I have bound shortcuts to the different fiiles in this harpoon list -
to get to the first file in this list I use <M-a> 
to get to the second <M-s>
to get to the third <M-q>
fourth <M-w>
fifth <M-z>
sixth <M-x>
```
		vim.keymap.set({ "n", "i", "v" }, "<M-a>", function()
			harpoon:list():select(1)
		end, { desc = "Navigate to file 1 in harpoon" })
		vim.keymap.set({ "n", "i", "v" }, "<M-s>", function()
			harpoon:list():select(2)
		end, { desc = "Navigate to file 2 in harpoon" })
		vim.keymap.set({ "n", "i", "v" }, "<M-q>", function()
			harpoon:list():select(3)
		end, { desc = "Navigate to file 3 in harpoon" })
		vim.keymap.set({ "n", "i", "v" }, "<M-w>", function()
			harpoon:list():select(4)
		end, { desc = "Navigate to file 4 in harpoon" })
		vim.keymap.set({ "n", "i", "v" }, "<M-z>", function()
			harpoon:list():select(5)
		end, { desc = "Navigate to file 5 in harpoon" })
		vim.keymap.set({ "n", "i", "v" }, "<M-x>", function()
			harpoon:list():select(6)
		end, { desc = "Navigate to file 6 in harpoon" })

```
where M is the alt key. 
What is the desired functionality that I want to achieve with the decorator:
I want to add a letter to the signcloumn next to each file signifying <alt + which key> i have to press to get to that file which i have started with the getharpoonlist function which now returns a table with {key : absolute path of file} -

I want to still keep the bookmark functionality but since I saw that #2948 replaced the bookmarks file I have first tried this. I also want to add a different highlight to eacg letter as well as file in tree - 
a - yellow
s - blue
q - purple
and so on, but that is an extra feature I had in mind not the core of my idea.

"It looks like getHarpoonList is only called once when the explorer is created and stored in a local, hence mappings will be out of date and incorrect
define_sign is called many times, creating or replacing a sign; it should only be done once. It's not clear how this should work and the expected results."

you are correct if there was a way to update the list manually i think you might be able to do it in the bufleave function which will update the decorator each time we leave the harpoon menu
```
local harpoon = require("harpoon")

harpoon:setup({
    ["my_list"] = {
        BufLeave = function(evt, list)
            -- evt is the event data; you can use it to determine the context
            -- list is the current Harpoon list

            print("Buffer left:", evt.bufnr)  -- Example action on buffer leave
            -- Perform any custom behavior here, such as saving state, logging, etc.
        end,
    },
})
```
The initial problem seems to be that you with the config I added while you see some letters next to the files that are a part of the harpoon list you would for example see for 
```
lua/myashkov/plugins/harpoon-lualine.lua
lua/myashkov/plugins/harpoon.lua
lua/myashkov/plugins/nvim-tree.lua
```
q lua/myashkov/plugins/harpoon-lualine.lua
q lua/myashkov/plugins/harpoon.lua
q lua/myashkov/plugins/nvim-tree.lua

and not 

a lua/myashkov/plugins/harpoon-lualine.lua
s lua/myashkov/plugins/harpoon.lua
q lua/myashkov/plugins/nvim-tree.lua

seems like even though we are doing calculateicons(node) we still update the self
```
if harpoon_letter then
    vim.notify("Harpoon letter: " .. harpoon_letter)
    self.icon = {
      str = harpoon_letter,
      hl = { "Search" },
    }
    self:define_sign(self.icon)
    return { self.icon }
  end
```
but if we try to return something other than self.icon or a tble identical to self.icon you just run into problems both with the hl group and the text.

**Comment by alex-courtis** [comment]
I attempted to replicate with this minimal config, however nvim-tree just shows an empty buffer on `:NvimTreeOpen` 
[nvt-min.lua.gz](https://github.com/user-attachments/files/17605057/nvt-min.lua.gz)

This all looks reasonable. It does look like using `self.icon =` replaces the icon for the one and only decorator instance. What happens when you return a new icon table for `HighlightedString[]` for `:calculate_icons`?

It's clear what additional functionality is required for the user/custom decorator. I'll make sure that's in there when it is implemented and there are clear examples.

I think it's best for you to wait for the custom decorators, as your use case extends far beyond the limited set of functionality that's available for the internal decorators.

**Comment by alex-courtis** [comment]
@DMyashkov please see #2948 

A UserDecorator proof of concept has been created, along with a minimal example to use. I would be most grateful for your testing.

**Comment by DMyashkov** [comment]
@alex-courtis Thank you so much, I will try to implement it as soon I can. I've been pretty busy with work lately but I'll update you when i create it.

**Reply by alex-courtis** [reply]
That would be fantastic!

**Reply by alex-courtis** [reply]
See work in progress wiki: https://github.com/nvim-tree/nvim-tree.lua/wiki/Decorators

Discussion on PR: https://github.com/nvim-tree/nvim-tree.lua/pull/2996

**Reply by alex-courtis** [reply]
Released 1.9.0
